src/lib.rs

```diff
@@ -127,12 +170,14 @@ impl StopSource {
 impl Future for StopToken {
     type Output = ();
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
-        let chan = Pin::new(&mut self.chan);
-        match Stream::poll_next(chan, cx) {
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(Some(never)) => match never {},
-            Poll::Ready(None) => Poll::Ready(()),
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        if let Some(mut listener) = self.signal.listen() {
```

Comment:
Calling `self.event.listen()` on each poll doesn't seem right to me: it allocates a new box, etc. 

Can we basically say that `type StopToken = Arc<EventListener>`?

Reply:
Note that I didn't have a chance to familiarize myseelf with `event_listener`, so there's a high probability of me being wrong here

Reply:
I agree that that seems unfortunate, especially if we can't in fact use the `Relaxed` ordering for this. I'll see if we can reuse the `EventListener` when I get home.

Reply:
After thinking about it a bit more I don't really know if that makes all that much sense though. Since Futures shouldn't be polled in a loop anyways, but should instead register a `Waker` it doesn't look like `poll` is usually called more than once anyways.
After some very sophisticated debugging (a println! in the `poll` function) it turns out that in the smoke test the future is polled once for each element in a `Stream` which makes sense if I think about it.

The problem with reusing the `EventListener` is that we need a mutable reference to it in order to poll it which would probably mean using a `Mutex` inside of the CondVar which I don't really like.

Reply:
I have managed to somewhat reduce the cost by  replacing the early creation of the `EventListener` with two checks of the `signaled` flag but I'll see if I can somehow manage to do this without just putting a Mutex in there. 

I'm thinking of atomically taking out the Pointer to the `EventListener` and creating a new `EventListener` as a fallback if the existing one isn't present or has not been initialized but that might be difficult without `unsafe` or a dependency on `crossbeam`.

Reply:
Okay I implemented the listener caching with a bit of `unsafe`.
I'm pretty sure about the safety but a second pair of eyes would be great.
Feel free to look over it when you have the time.

Reply:
Okay so I just found a double free bug. Apparently if we fail to insert a listener that we already made into a Box and turn it back into a `Box` it gets dropped twice. This is pretty weird since the only way to get that value is by a `SeqCst` atomic swap...

(relevant line: [lib.rs:157](https://github.com/async-rs/stop-token/pull/4/commits/6f4fab201c80f3126198a02d3c84118ba94f3d95#diff-b4aea3e418ccdb71239b96952d9cddb6R157))

Reply:
Okay I'm pretty sure it doesn't makes sense to cache listeners after all:
I can't think of a way to do it without a Mutex and that would kind of defeat the purpose of caching the listener.

Reply:
Never mind, I'm pretty sure I've got something that works now.